src/NCPkgFilterService.cc

```diff
@@ -0,0 +1,310 @@
+
+    setHeader( header);
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::addLine
+//	METHOD TYPE : void
+//
+//	DESCRIPTION : Add one line (with tag) to the service table
+//
+
+void NCPkgServiceTable::addLine ( ZyppService r, const std::vector <std::string> & cols )
+{
+    //use default ctor, add cell in the next step
+    YTableItem *tabItem = new YTableItem();
+
+    //place tag (with service reference) to the 0th column
+    tabItem->addCell( new NCPkgServiceTag ( r ) );
+
+    // and append the rest (name, URL and stuff)
+    for(const std::string& s: cols) {
+	tabItem->addCell(s);
+    };
+
+    // this is NCTable::addItem( tabItem );
+    //it actually appends the line to the table
+    addItem( tabItem );
+
+
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::getTag
+//	METHOD TYPE : NCPkgServiceTag *
+//
+//	DESCRIPTION : Get tag of service table line on current index,
+//		      (contains service reference)
+//
+
+NCPkgServiceTag* NCPkgServiceTable::getTag (const int & index )
+{
+   NCTableLine *line = myPad()->ModifyLine( index );
+   if ( !line )
+   {
+	return nullptr;
+   }
+
+   YTableItem *it = line->origItem();
+
+   //get actual service tag from 0th column of the table
+   YTableCell *tcell = it->cell(0);
+   NCPkgServiceTag *tag = static_cast<NCPkgServiceTag *>( tcell );
+
+   return tag;
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::getService
+//	METHOD TYPE : ZyppService
+//
+//	DESCRIPTION : Get service reference from selected line's tag
+//
+
+ZyppService NCPkgServiceTable::getService( int index )
+{
+    NCPkgServiceTag *t = getTag( index );
+
+    if ( !t )
+    {
+	return ZyppService( );
+    }
+    else
+    {
+	 return t->getService();
+    }
+}
+
+static std::string html_escape(const std::string& s)
+{
+    std::string escaped = boost::replace_all_copy(s, "&", "&amp;");
+    boost::replace_all(escaped, "<", "&lt;");
+    boost::replace_all(escaped, ">", "&gt;");
+    return escaped;
+}
+
+std::string NCPkgServiceTable::getDescription(ZyppService r)
+{
+    zypp::ServiceInfo si = repo_manager->getService(r);
+
+    std::string label = _( "<b>Service URL:</b>" );
+    std::string ret = label + html_escape(si.url().asString());
+    return ret;
+}
+
+/////////////////////////////////////////////////////////////////////
+////
+////
+////	METHOD NAME : NCPkgFilterService::fillServiceList
+////	METHOD TYPE : bool
+////
+////	DESCRIPTION : Add items to the service list (assoc.
+////		      product name, if any, and URL)
+////
+//
+bool NCPkgServiceTable::fillServiceList()
+{
+    yuiMilestone() << "Filling service list" << endl;
+
+    std::set<std::string> seen_services;
+
+    std::for_each(ZyppRepositoriesBegin(), ZyppRepositoriesEnd(), [&](const zypp::Repository& repo) {
```

Comment:
This will fill the services list in the order in which they appear in `ZyppRepositories`. I guess it won't be long until somebody will ask to have them sorted alphabetically or by some other sort order. When that happens, we'll have to put them into an array first, sort that array by that order and then interate over the array and add them to the widget one by one.